File: src/lib/state.ts
import { Annotation } from "@langchain/langgraph";
import {
  CompanyInfo,
  WebAnalysis,
  FinancialAnalysis,
  CompetitiveAnalysis,
  DimensionScores,
  SynthesisRationales,
} from "./types";

export const GraphState = Annotation.Root({
  companyInput: Annotation<string>(),

  companyInfo: Annotation<CompanyInfo>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),
  webAnalysis: Annotation<WebAnalysis>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),
  financialAnalysis: Annotation<FinancialAnalysis>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),
  competitiveAnalysis: Annotation<CompetitiveAnalysis>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),
  scores: Annotation<DimensionScores>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),
  synthesis: Annotation<SynthesisRationales>({
    reducer: (curr, update) => ({ ...curr, ...update }),
    default: () => ({}),
  }),

  verdict: Annotation<"INVEST" | "WATCH" | "PASS">(),
  confidence: Annotation<number>(),
  timeHorizon: Annotation<string>(),
  headline: Annotation<string>(),
  investThesis: Annotation<string>(),
  watchFor: Annotation<string[]>({
    reducer: (curr, update) => update, // overwrite
    default: () => [],
  }),
  comparableTo: Annotation<string>(),
  
  report: Annotation<string>(),

  streamLog: Annotation<string[]>({
    reducer: (curr, update) => [...curr, ...update],
    default: () => [],
  }),
  error: Annotation<string | null>(),
});

export type State = typeof GraphState.State;

File: backend/state.py
"""
state.py — LangGraph agent state
Port of src/lib/state.ts + src/lib/types.ts
"""

import operator
from typing import Annotated, Any, Optional
from typing_extensions import TypedDict


# ─── Reducers ─────────────────────────────────────────────────────────────────


def _merge_dicts(a: Optional[dict], b: Optional[dict]) -> Optional[dict]:
    """Shallow-merge two dicts; mirrors the TS `(curr, update) => ({...curr, ...update})` reducer."""
    if a is None:
        return b
    if b is None:
        return a
    return {**a, **b}


# ─── AgentState ───────────────────────────────────────────────────────────────


class AgentState(TypedDict, total=False):
    # ── Input ──────────────────────────────────────────────────────────────────
    companyInput: str

    # ── Node outputs ────────────────────────────────────────────────────────────
    # Merge reducers match the TS `(curr, update) => ({...curr, ...update})` pattern.
    companyInfo: Annotated[Optional[dict], _merge_dicts]
    webAnalysis: Annotated[Optional[dict], _merge_dicts]
    financialAnalysis: Annotated[Optional[dict], _merge_dicts]
    competitiveAnalysis: Annotated[Optional[dict], _merge_dicts]
    scores: Annotated[Optional[dict], _merge_dicts]
    synthesis: Annotated[Optional[dict], _merge_dicts]

    # ── Final decision ──────────────────────────────────────────────────────────
    verdict: Optional[str]            # "INVEST" | "WATCH" | "PASS"
    confidence: Optional[int]         # 50-99
    timeHorizon: Optional[str]
    headline: Optional[str]
    investThesis: Optional[str]
    watchFor: Optional[list]          # overwrite reducer (default last-write-wins)
    comparableTo: Optional[str]

    # ── Report ──────────────────────────────────────────────────────────────────
    report: Optional[str]             # full markdown report

    # ── Runtime ─────────────────────────────────────────────────────────────────
    # operator.add → append reducer (matches TS `[...curr, ...update]`)
    streamLog: Annotated[list[str], operator.add]
    error: Optional[str]
